File: validation/cases/lqi_tracking/lqi_tracking.m
% lqi_tracking.m -- LQI (LQR with integral action) step tracking.
% Mass-spring-damper with position output, tracks step reference.
%
% Usage: octave --no-gui lqi_tracking.m > lqi_tracking_octave.csv

pkg load control;

A_c = [0, 1; -1, -0.5];
B_c = [0; 1];
C_c = [1, 0];
D_c = [0];
dt = 0.05;
duration = 15.0;

sys_c = ss(A_c, B_c, C_c, D_c);
sys_d = c2d(sys_c, dt);
Ad = sys_d.A;
Bd = sys_d.B;
Cd = sys_d.C;

nx = 2;
ny = 1;
nu = 1;
nx_aug = nx + ny;

% Augmented system for integral action
A_aug = zeros(nx_aug);
A_aug(1:nx, 1:nx) = Ad;
A_aug(nx+1:end, 1:nx) = -Cd;
A_aug(nx+1:end, nx+1:end) = eye(ny);

B_aug = zeros(nx_aug, nu);
B_aug(1:nx, :) = Bd;

% Weight matrices for augmented system
Q_aug = diag([10, 1, 50]);  % position, velocity, integral
R_aug = [0.1];

K_aug = dlqr(A_aug, B_aug, Q_aug, R_aug);
Kx = K_aug(1:nu, 1:nx);
Ki = K_aug(1:nu, nx+1:end);

x = [0; 0];
xi = 0;  % integral state
ref = 1.0;

fprintf('time,position,velocity,integral,control\n');

t = 0.0;
while t < duration - dt/2
    y = Cd * x;
    u = -Kx * x - Ki * xi;

    fprintf('%.15e,%.15e,%.15e,%.15e,%.15e\n', t, x(1), x(2), xi, u);

    x = Ad * x + Bd * u;
    xi = xi + (ref - y);

    t = t + dt;
end

File: validation/cases/lqi_tracking/lqi_tracking.cpp
// lqi_tracking.cpp -- LQI tracking matching lqi_tracking.m
// Usage: ./lqi_tracking > lqi_tracking_cpp.csv

#include "ctrlpp/model/discretize.h"
#include "ctrlpp/model/propagate.h"
#include "ctrlpp/model/state_space.h"
#include "ctrlpp/control/lqr.h"

#include <cstdio>
#include <iostream>

int main()
{
    using Scalar = double;
    constexpr std::size_t NX = 2;
    constexpr std::size_t NU = 1;
    constexpr std::size_t NY = 1;

    ctrlpp::continuous_state_space<Scalar, NX, NU, NY> sys_c{};
    sys_c.A << 0.0, 1.0, -1.0, -0.5;
    sys_c.B << 0.0, 1.0;
    sys_c.C << 1.0, 0.0;
    sys_c.D << 0.0;

    constexpr Scalar dt = 0.05;
    constexpr Scalar duration = 15.0;

    auto sys_d = ctrlpp::discretize(ctrlpp::zoh{}, sys_c, dt);

    Eigen::Matrix<Scalar, 3, 3> Q_aug;
    Q_aug << 10.0, 0.0, 0.0,
             0.0, 1.0, 0.0,
             0.0, 0.0, 50.0;
    Eigen::Matrix<Scalar, 1, 1> R;
    R << 0.1;

    auto lqi_result = ctrlpp::lqi_gain<Scalar, NX, NU, NY>(sys_d.A, sys_d.B, sys_d.C, Q_aug, R);
    if(!lqi_result.has_value())
    {
        std::cerr << "LQI gain synthesis declined the validation plant\n";
        return 1;
    }
    auto lqi_res = *lqi_result;

    Eigen::Matrix<Scalar, 2, 1> x = Eigen::Matrix<Scalar, 2, 1>::Zero();
    Scalar xi = 0.0;
    constexpr Scalar ref = 1.0;

    std::printf("time,position,velocity,integral,control\n");

    for(Scalar t = 0.0; t < duration - dt / 2; t += dt)
    {
        Scalar y = (sys_d.C * x)(0);
        Eigen::Matrix<Scalar, 1, 1> u_vec = -lqi_res.Kx * x;
        Eigen::Matrix<Scalar, 1, 1> xi_vec;
        xi_vec << xi;
        u_vec -= lqi_res.Ki * xi_vec;
        Scalar u = u_vec(0);

        Eigen::Matrix<Scalar, 1, 1> u_input;
        u_input << u;

        std::printf("%.15e,%.15e,%.15e,%.15e,%.15e\n", t, x(0), x(1), xi, u);

        x = ctrlpp::propagate(sys_d, x, u_input);
        xi += ref - y;
    }
}
